### PR TITLE
Port fzf#shellescape to fix escaping issues in Windows

### PR DESCRIPTION
**Pull request by janlazo**
Initial work to fix all escaping issues for Vim and Neovim in Windows. `tempname` is necessary to bypass Vim's regular shellescaping in both `s:system` and `s:bang` like in Vim plugin of fzf because regular Vim seems to assume that the shell options are unchanged in Windows even if all but `shell` are unset. It's possible to not require batchfiles but it means double-escaping (ie. `fzf#shellescape(fzf#shellescape(cmd))` ) for Vim and single-escaping for Neovim with the same values for shell options.

### CODE REVIEW

File: plug.vim [hunk 1]
```diff
@@ -2316,10 +2345,19 @@ function! s:preview_commit()
...
-    let [sh, shrd] = s:chsh(1)
-    execute 'silent %!cd' s:shellesc(g:plugs[name].dir) '&& git show --no-color --pretty=medium' sha
+    let [sh, shellcmdflag, shrd] = s:chsh(1)
+    let cmd = 'cd '.s:shellesc(g:plugs[name].dir).' && git show --no-color --pretty=medium '.sha
+    if s:is_win
+      let batchfile = tempname().'.bat'
+      call writefile(['@echo off', cmd], batchfile)
+      let cmd = batchfile
+    endif
+    execute 'silent %!' cmd
```
Comment (janlazo) [inline plug.vim]: This looks like `s:bang` now but `%#!` are not escaped.

**Comment by janlazo** [comment]
`s:spawn` needs `tempname` for the jobs for the same reasons on `system` and `:!`, especially in regular Vim 8.

**Comment by janlazo** [comment]
@junegunn Do I have to update the python and ruby jobs?

**Comment by junegunn** [comment]
> Do I have to update the python and ruby jobs?

No, we don't want to spend more of our time on those legacy installers.

So this fixes #606, right? It's not clear to me though how using a batchfile instead of a command string suppresses "Enter key" prompt. Can you explain?

**Comment by janlazo** [comment]
This targets #635, #668, #539 ('shell' points to `$SHELL`), Case 1 in https://github.com/junegunn/vim-plug/wiki/faq#windows-system-error-e484.

For 606, appending `silent` to the bang command in `s:bang` (see https://github.com/junegunn/vim-plug/issues/606#issuecomment-326469010) is enough. I don't know if this is required for MacVim as well.

**Comment by junegunn** [comment]
> This PR targets #635, #668, #539 

Thanks for the clarification. I completely lost track of those issues, largely because I'm not a Windows user.

I just reviewed the code, although it's unfortunate that we have to introduce extra code branches here and there for Windows, the change looks good to me. Is there anything else do you want to address?

**Comment by janlazo** [comment]
>  Can you update the commit message with `Close #xxx` lines so that those issues are automatically closed when this is merged?

I'll rebase when I think it's ready.

> Is there anything else you want to address?

I prefer vim-plug to use `cmd.exe` only in Windows so the side effects of setting `SHELL` before Vim runs should be reverted for `system` to work properly.

**Comment by janlazo** [comment]
@junegunn do you want the fix for #606 in here or in a separate PR?

**Comment by janlazo** [comment]
@junegunn Unlike fzf's Vim plugin, `shell` is not set at the start so `s:shellesc` may not work as expected. To guarantee that it will work in the batchfile, I think it's better to make it non-configurable.

```viml
if has('win32')
  function s:shellesc(arg)
    return s:shellesc_cmd(a:arg)
  endfunction
else
  function s:shellesc(arg)
    return shellescape(a:arg)
  endfunction
endif
```

**Comment by junegunn** [comment]
> do you want the fix for #606 in here or in a separate PR?

As long as it's on a separate commit, I'm fine with either :) By the way, we are going to apply `silent` only on GVim, right? As you did in https://github.com/junegunn/vim-plug/issues/606#issuecomment-326469010. On terminal Vim, it messes up the screen and requires additional `:redraw!`.

> do you want to keep the old s:shellesc as fallback instead of shellescape?

No, you properly handled escaping issues on Windows, so I guess that crude solution is no longer needed.

**Comment by junegunn** [comment]
Reviewed the code and it looks good to me. Thanks, again. I see that you rebased your commits, ready for merge?

**Comment by janlazo** [comment]
> As long as it's on a separate commit, I'm fine with either :) By the way, we are going to apply `silent` only on GVim, right?

Yes. I'll open a separate PR for #606 after the merge.

> ready for merge?

Yes for Windows unless you want to handle msysgit or cygwin as well.

**Comment by junegunn** [comment]
Thanks, I really appreciate your contribution.

**Comment by janlazo** [comment]
@junegunn Do you know why some `s:system` calls use `s:esc` for tags, branches instead of `s:shellesc`?

I got https://github.com/junegunn/vim-plug/commit/eb47183af2374f278b776f369c8cecb3c9441f49 from the blame. I don't use tags or branches so I didn't notice while working on this PR.

**Pull request merged**